File: 107. Binary Tree Level Order Traversal II/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function(root) {
  if (!root) return [];

  const result = [];
  const queue = [root];
  while (queue.length > 0) {
    const current = [];
    const size = queue.length;

    for (let i = 0; i < size; i++) {
      const head = queue.shift();
      current.push(head.val);
      if (head.left) queue.push(head.left);
      if (head.right) queue.push(head.right);
    }

    result.unshift(current);
  }

  return result;
};

File: 107. Binary Tree Level Order Traversal II/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func bfs(s *[][]int, level int, root *TreeNode) {
	if root == nil {
		return
	}
	if level >= len(*s) {
		*s = append(*s, []int{})
	}
	(*s)[level] = append((*s)[level], root.Val)
	bfs(s, level+1, root.Left)
	bfs(s, level+1, root.Right)
}

func levelOrderBottom(root *TreeNode) [][]int {
	var s [][]int
	bfs(&s, 0, root)
	for i := 0; i < len(s)/2; i++ {
		s[i], s[len(s)-i-1] = s[len(s)-i-1], s[i]
	}
	return s
}